{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3b2caf9f-3b69-42f8-b9cd-58fe4beb40bd",
   "metadata": {},
   "source": [
    "# Baltic Sea BioMed summer school: PROJECT WORK, DAY 2 \n",
    "This notebook contains analysis of RNA-seq data generated from human liver cancer cell line **[HEPG2](https://en.wikipedia.org/wiki/Hep_G2)**, that is commonly used model for liver toxicity. Cells have been treated with DMSO (control), PCB118 or PCB153 for 24 hours. PCB118 and PCB153 are **[polychlorinated biphenyls](https://en.wikipedia.org/wiki/Polychlorinated_biphenyl)**. RNA-seq libraries are produced from mRNAs, mapped to human genome hg38 and differential gene expression (DGE) is analyzed with **[DESeq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8)** and using DMSO as a reference.  \n",
    "\n",
    "  **DAY 1:** We'll get familiar with the data and produce various common visualizations.  \n",
    "  **DAY 2:** We'll do gene-set enrichment analysis (GSEA) and predict upstream transcription regulators using Lisa online tool. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253a52a6-5af2-4797-a308-a4de9d2fd2f0",
   "metadata": {},
   "source": [
    "## 1. Load libraries and import RNA-seq data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6e02417-9a44-4ae6-8767-a2d15aaa0bbe",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## load libraries\n",
    "# these are used in GSEA analysis\n",
    "library(clusterProfiler)\n",
    "library(enrichplot)\n",
    "\n",
    "# these are used for plotting\n",
    "library(ggplot2)\n",
    "library(ggrepel)\n",
    "library(gridExtra)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41123611-1683-4d01-8547-2ad2eafe598e",
   "metadata": {},
   "source": [
    "Next, we'll use `readRDS()` to read-in the data from yesterday (remember that we stored it as R object?). We also use `set.seed()` to set seed value for random number generators. This is done to retain consistent results. You can choose your own seed number :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8de391f-96d4-40b3-baad-10928f70e472",
   "metadata": {},
   "outputs": [],
   "source": [
    "## read RNA-seq data frame from day 1\n",
    "# define home directory\n",
    "home_dir <- getwd()\n",
    "\n",
    "# read saved data frame as df and check that it's ok\n",
    "df <- readRDS(paste0(home_dir, \"/df.rds\"))\n",
    "head(df)\n",
    "\n",
    "# set random generator seed for consistent results (you can choose your own)\n",
    "set.seed(38642)\n",
    "\n",
    "# TPM cutoff might be needed \n",
    "cutoff_tpm <- 1   # cutoff value for expressed genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5512e3b7-2386-4a4e-902c-1cb620a040d9",
   "metadata": {},
   "source": [
    "## 2. Gene-set enrichment analysis (GSEA)\n",
    "More comprehensive instructions and theoretical background can be found at **[Biomedical knowledge mining book](https://yulab-smu.top/biomedical-knowledge-mining-book/index.html)**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "559659a6-9268-48c1-8b9b-37d2ab792c71",
   "metadata": {},
   "source": [
    "**OPTIONAL STEP:** download Hallmark gene set from mSigDB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b74d19bc-35be-4744-9512-fc363553a8a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## OPTIONAL: load pathways from mSigDB ##\n",
    "#########################################\n",
    "## you can browse mSigDB signature pathways from their webpage:\n",
    "# https://www.gsea-msigdb.org/gsea/msigdb/index.jsp\n",
    "\n",
    "## or using msigdbr tool:\n",
    "library(msigdbr) # MSigDB - molecular signature data base\n",
    "as.data.frame(msigdbr_collections()) # check pathway categories and subcategories. \n",
    "\n",
    "# load gene set names and gene symbols belonging to hallmark category\n",
    "pathway <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"H\", \n",
    "                    subcategory = \"\")\n",
    "\n",
    "# An other example to download reactome pathways\n",
    "# pathway <- msigdbr(species = \"Homo sapiens\", category = \"C2\", subcategory = \"CP:REACTOME\")\n",
    "\n",
    "# this downloads table as a \"tibble\" format with various gene annotations\n",
    "pathway\n",
    "\n",
    "# for GSEA, we only need columns for gene set name (gs_name) and associated gene symbols (gene_symbols)\n",
    "pathway <- pathway[,c(\"gs_name\", \"gene_symbol\")]\n",
    "pathway\n",
    "\n",
    "# write pathway to file for later use\n",
    "write.table(x = pathway, \n",
    "           file = paste0(home_dir,\"/Hallmark_pathway_mSigDB.tsv\"),\n",
    "           quote = FALSE,\n",
    "           row.names = FALSE,\n",
    "           col.names = TRUE,\n",
    "           sep = \"\\t\"\n",
    "           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00e09429-95a3-4e18-a8e9-1aeb9e3ce320",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## GSEA of PCB118 treatment responses\n",
    "# save DMSO vs PCB118 logFC values as a matrix and add gene symbols as rownames\n",
    "ma2 <- as.matrix(df$DMSO.50nM_vs_PCB118.50microM_Log2_Fold_Change)\n",
    "rownames(ma2) <- df$symbol\n",
    "colnames(ma2) <- \"PCB118\"\n",
    "\n",
    "# order according to logFC \n",
    "ma2 <- ma2[order(ma2, decreasing = TRUE),]\n",
    "\n",
    "# check if there are NAs\n",
    "table(is.na(ma2))\n",
    "\n",
    "# remove NAs\n",
    "ma2 <- na.omit(ma2)\n",
    "\n",
    "# check first and last entries\n",
    "head(ma2)\n",
    "tail(ma2)\n",
    "\n",
    "# upload hallmarks pathway (if not done already)\n",
    "pathway <- read.delim(paste0(home_dir,\"/Hallmark_pathway_mSigDB.tsv\"))\n",
    "\n",
    "## run GSEA analysis (set pvalueCutoff to 1 if you want to have all results)\n",
    "gsea_results <- GSEA(geneList = ma2, \n",
    "                     TERM2GENE = pathway, \n",
    "                     pvalueCutoff = 0.05)\n",
    "\n",
    "# you can look results as a data frame \n",
    "gsea_table <- as.data.frame(gsea_results)\n",
    "gsea_table[, 1:(ncol(gsea_table)-2)] # last columns are a bit too big for screen.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "343e4b54-4e9b-4982-943d-f29d0a2b2cdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# classical GSEA plots\n",
    "gseaplot(gsea_results, by = \"all\", title = gsea_results$Description[1], geneSetID = 1)\n",
    "#gseaplot(gsea_results, by = \"all\", title = gsea_results$Description[2], geneSetID = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eb12b40-2c11-48f7-b577-bff1cb5d572e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## dotplot (from enrichplot package)\n",
    "dotplot(gsea_results, showCategory=15, split=\".sign\") + facet_grid(.~.sign)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72f2f005-7d21-46e0-8ade-76cd452148f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## network plot for GSEA results\n",
    "# select how many GSEA categories to plot\n",
    "top_n_category <- 5\n",
    "\n",
    "# plot network graph showing shared genes and fold change from matrix as color\n",
    "#pdf(paste0(home_dir, \"cnet_plot_of_top_\", top_n_category ,\"_GSEA_categories.pdf\"))\n",
    "cnetplot(gsea_results, \n",
    "         categorySize = \"geneNum\", \n",
    "         foldChange = ma2, \n",
    "         showCategory = top_n_category, \n",
    "         node_label = \"category\")\n",
    "#dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08fec455-4b1c-4c70-ac9d-8c320659e329",
   "metadata": {},
   "outputs": [],
   "source": [
    "## PCB153 GSEA analysis\n",
    "# copy and modify the above code for this"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4021f8f-742b-48fe-939f-4bb2553160dd",
   "metadata": {},
   "source": [
    "## 3. Upstream regulator prediction using Lisa\n",
    "Next we'll use online tool **Lisa** **[(documentation)](http://lisa.cistrome.org/doc)**, to predict upstream transcription regulators. With this tool we can predict which transcription factors and coregulators are responsible for inducing and repressed genes in response to PCB118 or PCB153 treatments. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5eefbef-cb34-45fc-9f98-fa24b9075742",
   "metadata": {},
   "source": [
    "### 3.1 Prepare gene lists \n",
    "Lisa want's to have 50-500 genes to make reliable predictions. Our DEG categories are all bigger than 500 genes. We'll order genes based on adjusted p-value and take top 500 induced (up) or repressed (dn) genes for our gene lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63e4e943-f0d8-4f12-9829-f0c10e71072e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## write table of induced gene symbols for LISA analysis: http://lisa.cistrome.org/\n",
    "\n",
    "# create directory for the datan",
    "dir.create(paste0(home_dir, \"/data"),\n",)",
    "\n",
    "# sort according to adj. pvalue for PCB118 analysis\n",
    "df <- df[order(df$DMSO.50nM_vs_PCB118.50microM_adj_pvalue),]\n",
    "\n",
    "# write table with top 500 PCB118-induced gene symbols for LISA analysis\n",
    "write.table(x = head(df[df$PCB118_up == 1, ]$symbol, n = 500),\n",
    "          file = paste0(home_dir, \"/data/PCB118_up_top500.txt\"),\n",
    "          quote = FALSE, row.names = FALSE, col.names = FALSE)\n",
    "          \n",
    "# write table with top 500 PCB118-induced gene symbols for LISA analysis\n",
    "write.table(x = head(df[df$PCB118_dn == 1, ]$symbol, n = 500),\n",
    "          file = paste0(home_dir, \"/data/PCB118_dn_top500.txt\"),\n",
    "          quote = FALSE, row.names = FALSE, col.names = FALSE)\n",
    "\n",
    "# sort according to adj. pvalue for PCB153 analysis\n",
    "df <- df[order(df$DMSO.50nM_vs_PCB153.50microM_adj_pvalue),]\n",
    "\n",
    "# write table with top 500 PCB153-induced gene symbols for LISA analysis\n",
    "write.table(head(df[df$PCB153_up == 1, ]$symbol, n = 500),\n",
    "          file = paste0(home_dir, \"/data/PCB153_up_top500.txt\"),\n",
    "          quote = FALSE, row.names = FALSE, col.names = FALSE)\n",
    "          \n",
    "# write table with top 500 PCB153-induced gene symbols for LISA analysis\n",
    "write.table(x = head(df[df$PCB153_dn == 1, ]$symbol, n = 500),\n",
    "          file = paste0(home_dir, \"/data/PCB153_dn_top500.txt\"),\n",
    "          quote = FALSE, row.names = FALSE, col.names = FALSE)\n",
    "\n",
    "# continue with these gene lists on Lisa online tool\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce8075d0-dc45-4376-88ef-d369352fe0d8",
   "metadata": {},
   "source": [
    "### 3.2 run Lisa online tool\n",
    "Open **[Lisa](http://lisa.cistrome.org/)** online tool. Select **Species: Human** and insert gene symbols (copy-paste from text editor) of induced (up) genes to **Gene set 1**, and repressed (dn) gene lists to **Gene set 2**. Add **Job name: PCB118 or PCB153** and optionally email address. Click **Run** to initiate analysis. If you added email address, you will be notified when the run has been completed.  \n",
    "Once the runs are complete, press **Dowload all samples**, dowload and unzip the results .zip file to **home_dir/lisa_results/**.  \n",
    "Analysis produces ~60 files, but we'll only use two that contain de-duplicated transcription regulatory predictions based on ChIP-seq data for induced (gene set 1 = gs1) and repressed (gs2) genes. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05056db9-cf7a-42c6-b204-20a902cc5937",
   "metadata": {},
   "source": [
    "### 3.3 Prepare figures from Lisa results\n",
    "Let's build a similar scatter-plot of lisa results as is shown on Lisa webpage and label top-n most significant transcription factors (TFs)!  \n",
    "\n",
    "This time well make a vector of file names that contain \"chipseq_cauchy_combine_dedup\" and use `lapply()` function (\"list apply\") to read in result files as a list. Scatter plot images of -log(p-values) are generated in `lapply()` loop and lables are added to top_n transcription regulators with lowest p-value. Images are collected to a list called p3, which can be accessed later on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e525183-e4a4-48ac-ad14-9d0a1b734e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simple scatter plot\n",
    "## define how many TFs you would like to label in plot\n",
    "top_n <- 20\n",
    "\n",
    "## upload Lisa results\n",
    "# set home directory to point project folder with Lisa results\n",
    "lisa_res_dir <- paste0(home_dir, \"/lisa_results/\")\n",
    "\n",
    "# select files with results from ChIP-seq model\n",
    "in_files <- list.files(lisa_res_dir, pattern = \"chipseq_cauchy_combine_dedup\", full.names = TRUE)\n",
    "# in_files\n",
    "\n",
    "# if you followed instructions while uploading data to Lisa, your...\n",
    "# predictions for up genes have gs1 in file name, eg. PCB153_2024_07_12_0739250.215__hg38_gs1.txt_chipseq_cauchy_combine_dedup.csv\n",
    "# predictions for dn genes have gs2 in file name, eg. PCB153_2024_07_12_0739250.215__hg38_gs2.txt_chipseq_cauchy_combine_dedup.csv\n",
    "\n",
    "# read in data from the files as a list\n",
    "lisa_results <- lapply(in_files, function(i){\n",
    "  read.csv(i)\n",
    "})\n",
    "\n",
    "# name list items (check that the naming corresponds the order of your files!)\n",
    "names(lisa_results) <- c(\"PCB118_up\", \"PCB118_dn\", \"PCB153_up\", \"PCB153_dn\")\n",
    "\n",
    "# i=1\n",
    "# collect results from different treatments to a list\n",
    "p3 <- lapply(1:(length(lisa_results)/2), function(i){\n",
    "  ## predictions for repressed genes\n",
    "  # i*2 is the dn group and i*2-1 is the up group\n",
    "  dn_data <- lisa_results[[i*2]]\n",
    "  # rename p-value column to contain \"dn_\" prefix\n",
    "  names(dn_data)[2] <- paste0(\"dn_\", names(dn_data)[2])\n",
    "  \n",
    "  ## predictions for induced genes\n",
    "  up_data <- lisa_results[[i*2-1]]\n",
    "  # rename p-value column to contain \"up_\" prefix\n",
    "  names(up_data)[2] <- paste0(\"up_\", names(up_data)[2])\n",
    "  \n",
    "  # check that both up and dn data contain same TFs before merging data frames \n",
    "  print(paste0(\"dn and up data contain the same TFs? \", identical(sort(dn_data$TF), sort(up_data$TF))))\n",
    "  # not same TF data is kept for both though\n",
    "  # identical(sort(dn_data$X), sort(up_data$X))\n",
    "  \n",
    "  # merge data frames for plotting\n",
    "  data <- merge.data.frame(x = up_data, y = dn_data, by = \"TF\")\n",
    "  \n",
    "  # add negative log10 p-values  for plotting\n",
    "  data$neg_log_pval_up <- -log10(data$up_pval)\n",
    "  data$neg_log_pval_dn <- -log10(data$dn_pval)\n",
    "  \n",
    "  # add rank data\n",
    "  data$rank_up <- rank(data$up_pval)\n",
    "  data$rank_dn <- rank(data$dn_pval)\n",
    "  data$is_top_n <- data$rank_up < top_n | data$rank_dn < top_n\n",
    "  data$label <- ifelse(data$is_top_n, data$TF, \"\")\n",
    "  \n",
    "  # treatment name\n",
    "  treatment_name <- gsub(\"_dn\", \"\", names(lisa_results[i*2]))\n",
    "  \n",
    "  # plot \n",
    "  ggplot(data, aes(x=neg_log_pval_up, y = neg_log_pval_dn, label = label))+\n",
    "    theme_classic()+\n",
    "    geom_point(shape = 19, col=rgb(1,0,0,0.5))+\n",
    "    geom_text_repel(max.overlaps = 50, size = 4, color = \"black\")+\n",
    "    labs(title = paste0(\"Lisa predictions for \", treatment_name, \" treatment (top \",top_n,\" TFs labeled)\"),\n",
    "         x = paste0(names(lisa_results[i*2 -1]), \"  [-log10(p-value)]\"),\n",
    "         y = paste0(names(lisa_results[i*2]), \" [-log10(pvalue)]\"))\n",
    "  \n",
    "})\n",
    "\n",
    "# images were colletecd to a list named p3.\n",
    "# show first image\n",
    "p3[1]\n",
    "\n",
    "# show second image\n",
    "p3[2]\n",
    "\n",
    "## show/print both images in a grid\n",
    "# uncomment \"pdf()\" and \"dev.off()\" lines to print pdf of the results\n",
    "# pdf(paste0(home_dir, \"/Lisa_results_simple_top\", top_n, \".pdf\"), height = 16, width = 8)\n",
    "# do.call(grid.arrange, p3)\n",
    "# dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "306e3b40-0028-4b91-b044-be3cf3edadb4",
   "metadata": {},
   "source": [
    "Next we'll generate a bit more detailed scatter plot that combines Lisa results data with the RNA-seq information.  \n",
    "Here we'll combine two data frames using `merge.data.frame()` function. Then we'll remove Lisa predictions that are not expressed in the RNA-seq data (ie. do not meet TPM cutoff value), adjust the circle sizes to represent median expression in RNA-seq, and add logFC from the RNA-seq data as a color. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc8d02d2-30f6-46bc-a744-890d6bf57884",
   "metadata": {},
   "outputs": [],
   "source": [
    "# a more detailed scatter plot with expression data from RNA-seq\n",
    "## define how many TFs you would like to label in plot\n",
    "top_n <- 20\n",
    "\n",
    "## upload Lisa results\n",
    "# set home directory to point project folder with Lisa results\n",
    "lisa_res_dir <- paste0(home_dir, \"/lisa_results/\")\n",
    "\n",
    "# select files with results from ChIP-seq model\n",
    "in_files <- list.files(lisa_res_dir, pattern = \"chipseq_cauchy_combine_dedup\", full.names = TRUE)\n",
    "# in_files\n",
    "\n",
    "# if you followed instructions while uploading data to Lisa, your...\n",
    "# predictions for up genes have gs1 in file name, eg. PCB153_2024_07_12_0739250.215__hg38_gs1.txt_chipseq_cauchy_combine_dedup.csv\n",
    "# predictions for dn genes have gs2 in file name, eg. PCB153_2024_07_12_0739250.215__hg38_gs2.txt_chipseq_cauchy_combine_dedup.csv\n",
    "\n",
    "# read in data from the files as a list\n",
    "lisa_results <- lapply(in_files, function(i){\n",
    "  read.csv(i)\n",
    "})\n",
    "\n",
    "# name list items (check that the naming corresponds the order of your files!)\n",
    "names(lisa_results) <- c(\"PCB118_up\", \"PCB118_dn\", \"PCB153_up\", \"PCB153_dn\")\n",
    "\n",
    "# collect results from different treatments to a list\n",
    "p4 <- lapply(1:(length(lisa_results)/2), function(i){\n",
    "  ## predictions for repressed genes\n",
    "  # i*2 is the dn group and i*2-1 is the up group\n",
    "  dn_data <- lisa_results[[i*2]]\n",
    "  # rename p-value column to contain \"dn_\" prefix\n",
    "  names(dn_data)[2] <- paste0(\"dn_\", names(dn_data)[2])\n",
    "  \n",
    "  ## predictions for induced genes\n",
    "  up_data <- lisa_results[[i*2-1]]\n",
    "  # rename p-value column to contain \"up_\" prefix\n",
    "  names(up_data)[2] <- paste0(\"up_\", names(up_data)[2])\n",
    "  \n",
    "  # check that both up and dn data contain same TFs before merging data frames \n",
    "  print(paste0(\"dn and up data contain the same TFs? \", identical(sort(dn_data$TF), sort(up_data$TF))))\n",
    "  # not same TF data is kept for both though\n",
    "  # identical(sort(dn_data$X), sort(up_data$X))\n",
    "  \n",
    "  # merge data frames for plotting\n",
    "  data <- merge.data.frame(x = up_data, y = dn_data, by = \"TF\")\n",
    "  \n",
    "  # merge expression data from df\n",
    "  data <- merge.data.frame(x = data, y = df, by.x = \"TF\", by.y = \"symbol\")\n",
    "  \n",
    "  # remove non-expressed\n",
    "  data <- data[data$is_exp,]\n",
    "  \n",
    "  # add negative log10 p-values  for plotting\n",
    "  data$neg_log_pval_up <- -log10(data$up_pval)\n",
    "  data$neg_log_pval_dn <- -log10(data$dn_pval)\n",
    "  \n",
    "  # add rank data\n",
    "  data$rank_up <- rank(data$up_pval)\n",
    "  data$rank_dn <- rank(data$dn_pval)\n",
    "  data$is_top_n <- data$rank_up < top_n | data$rank_dn < top_n\n",
    "  data$label <- ifelse(data$is_top_n, data$TF, \"\")\n",
    "  \n",
    "  # treatment name\n",
    "  treatment_name <- gsub(\"_dn\", \"\", names(lisa_results[i*2]))\n",
    "  \n",
    "  # factorize expression values for size plotting\n",
    "  data$tpm_fact <- cut(data$median_exp, \n",
    "                       right = FALSE, # right border open-ended: [1,2)\n",
    "                       breaks = c(0, 2.5, 5, 10, 20, 30, 1000), \n",
    "                       labels = c(\"<2.5\", \"<5\", \"<10\", \"<20\", \"<30\",\"30-\"),\n",
    "                       ordered = TRUE)\n",
    "  # grep fold change column name for later use\n",
    "  col_name_logfc <- grep(\"Log2_Fold_Change\", grep(treatment_name, names(df), value = T), value = T)\n",
    "  \n",
    "  # plot \n",
    "  ggplot(data, aes(x=neg_log_pval_up, \n",
    "                   y = neg_log_pval_dn, \n",
    "                   label = label, \n",
    "                   size = tpm_fact,\n",
    "                   # col = grep(\"Log2_Fold_Change\", grep(treatment_name, names(df), value = T), value = T)))+\n",
    "                   # col = col_name_logfc))+\n",
    "                   col = .data[[col_name_logfc]]))+\n",
    "    theme_classic()+\n",
    "    geom_point()+\n",
    "    scale_color_gradientn(colors = rev(c(\"#A50026\", \"#F46D43\", \"#FDAE61\", \"#FFFFBF\", \"#E0F3F8\",  \"#74ADD1\", \"#313695\")),\n",
    "                          values = scales::rescale(c(min(data[,col_name_logfc], na.rm = TRUE), -1, -0.5, 0, 0.5 , 1, max(data[,col_name_logfc], na.rm = TRUE))),\n",
    "                          na.value = \"gray\")+\n",
    "    # \n",
    "    # scale_size_discrete(range = c(2,6), na.value = 1)+\n",
    "    # scale_size_continuous(trans = \"log10\", range = c(1, 10)) + \n",
    "    geom_text_repel(max.overlaps = 50, size = 4, color = \"black\")+\n",
    "    labs(title = paste0(\"Lisa predictions for \", treatment_name, \" treatment (top \",top_n,\" expressed TFs labeled). TPM > \", cutoff_tpm, \".\"),\n",
    "         x = paste0(names(lisa_results[i*2 -1]), \"  [-log10(p-value)]\"),\n",
    "         y = paste0(names(lisa_results[i*2]), \" [-log10(pvalue)]\"),\n",
    "         col = paste0(treatment_name, \" vs DMSO\\n(Log2FC)\"),\n",
    "         size = \"Expression\\n(median TPM)\")\n",
    "  \n",
    "})\n",
    "\n",
    "# show 1st image\n",
    "p4[1]\n",
    "\n",
    "# show 2nd image\n",
    "p4[2]\n",
    "\n",
    "## show ggplot images in grid\n",
    "# uncomment \"pdf()\" and \"dev.off()\" lines to print pdf of the results\n",
    "# pdf(paste0(home_dir, \"/Lisa_results_pretty_top\", top_n, \".pdf\"), height = 16, width = 8)\n",
    "# do.call(grid.arrange, p4)\n",
    "# dev.off()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ea8d0cb-363e-4836-9cea-ebd23699a9a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
